cpp/inc/bond/core/nullable.h

```diff
@@ -482,32 +474,37 @@ class nullable<T, Allocator, false>
 #endif
 
 private:
-    template<typename AllocatorT>
-    void destroy(AllocatorT& alloc)
-    {
 #ifndef BOND_NO_CXX11_ALLOCATOR
-        std::allocator_traits<AllocatorT>::destroy(alloc, _value);
+    typedef typename std::allocator_traits<allocator_type>::
+        template rebind_alloc<value_type> rebind_alloc;
+
+    typedef typename std::allocator_traits<rebind_alloc>::pointer real_pointer;
 #else
-        _value->~T();
+    typedef typename allocator_type::template rebind<value_type>::other rebind_alloc;
+
+    typedef pointer real_pointer;
 #endif
-        alloc.deallocate(_value, 1);
-    }
 
-    void destroy(detail::no_allocator&)
+    void delete_value()
     {
-        delete _value;
+#ifndef BOND_NO_CXX11_ALLOCATOR
+        std::allocator_traits<rebind_alloc>::destroy(_alloc, _value);
+#else
+        _value->~T();
+#endif
+        _alloc.deallocate(_value, 1);
     }
 
 #ifndef BOND_NO_CXX11_RVALUE_REFERENCES
     template<typename AllocatorT, typename Arg1>
-    pointer new_value(AllocatorT& alloc, Arg1&& arg1)
+    real_pointer new_value(AllocatorT& alloc, Arg1&& arg1)
     {
-        T* p = alloc.allocate(1);
+        real_pointer p(alloc.allocate(1));
         try
         {
 #ifndef BOND_NO_CXX11_ALLOCATOR
-            std::allocator_traits<AllocatorT>::construct(alloc, p,
-                std::forward<Arg1>(arg1));
+            std::allocator_traits<AllocatorT>::construct(alloc,
+                boost::addressof(*p), std::forward<Arg1>(arg1));
```

Comment:
You're on the right track. The type of `p` could be some smart pointer `X`, but `X` does not have to be convertible to type `U*` (for template parameter `U`) which is what `construct` takes as of C++11. The type of `&*p` could be some smart pointer `Y`, but `Y` also does not have to be convertible to the aforementioned `U*`.

The idiomatic thing, which is what I do here, which is also what C++ standard library allocator-aware containers do, which is also what you'll find in any correct allocator-aware code (e.g. in Boost.Container) is: Construct in the true address obtained by `addressof(*p)`.

A similar rationale governs the use of `::new((void*)p) Type...` instead of just `new(p) Type...`. Even if there exists a placement new overload that the latter would select, we don't care, in favor of choosing the most true placement new (for `void*` and in the global scope).